From 18532c2e13efbc1aa43f0b00b4459831f9414de0 Mon Sep 17 00:00:00 2001
From: SciLor <scilor@users.noreply.github.com>
Date: Mon, 19 Oct 2026 12:00:00 +0000
Subject: [PATCH] d3d9: reject cube/unsupported native textures safely

Cube-map native textures and failed allocations/locks return nil instead of asserting.
Authors: Dryxio (Southland-FR/librw 4406ffa, e91821e); same cube fix as aap/librw PR #161 (IgorPolyakov 9f6b37d). MIT.

Upstream-ref: https://github.com/Southland-FR/librw/commit/4406ffa and https://github.com/Southland-FR/librw/commit/e91821e
---
 src/d3d/d3d9.cpp | 39 +++++++++++++++++++++++++++++++--------
 1 file changed, 31 insertions(+), 8 deletions(-)

diff --git a/src/d3d/d3d9.cpp b/src/d3d/d3d9.cpp
index 851397b..79846a6 100644
--- a/src/d3d/d3d9.cpp
+++ b/src/d3d/d3d9.cpp
@@ -741,11 +741,19 @@ readNativeTexture(Stream *stream)
 	Raster *raster;
 	D3dRaster *ext;
 
+	if(flags & 2){
+		RWERROR((ERR_GENERAL, "D3D9 cube maps are not supported"));
+		tex->destroy();
+		return nil;
+	}
+
 	if(flags & 8){
 		// is compressed
-		assert((flags & 2) == 0 && "Can't have cube maps yet");
 		raster = Raster::create(width, height, depth, format | type | Raster::DONTALLOCATE, PLATFORM_D3D9);
-		assert(raster);
+		if(raster == nil){
+			tex->destroy();
+			return nil;
+		}
 		ext = GETD3DRASTEREXT(raster);
 		ext->format = d3dformat;
 		ext->hasAlpha = flags & 1;
@@ -753,14 +761,19 @@ readNativeTexture(Stream *stream)
 		                             raster->format & Raster::MIPMAP ? numLevels : 1,
 		                             0,
 		                             ext->format);
-		assert(ext->texture);
+		if(ext->texture == nil){
+			tex->raster = raster;
+			tex->destroy();
+			return nil;
+		}
 		raster->flags &= ~Raster::DONTALLOCATE;
 		ext->customFormat = 1;
-	}else if(flags & 2){
-		assert(0 && "Can't have cube maps yet");
 	}else{
 		raster = Raster::create(width, height, depth, format | type, PLATFORM_D3D9);
-		assert(raster);
+		if(raster == nil){
+			tex->destroy();
+			return nil;
+		}
 		ext = GETD3DRASTEREXT(raster);
 	}
 	tex->raster = raster;
@@ -774,15 +787,25 @@ readNativeTexture(Stream *stream)
 
 	uint32 size;
 	uint8 *data;
+	bool uploadOk = true;
 	for(int32 i = 0; i < numLevels; i++){
 		size = stream->readU32();
 		if(i < raster->getNumLevels()){
 			data = raster->lock(i, Raster::LOCKWRITE|Raster::LOCKNOFETCH);
-			stream->read8(data, size);
-			raster->unlock(i);
+			if(data){
+				stream->read8(data, size);
+				raster->unlock(i);
+			}else{
+				stream->seek(size);
+				uploadOk = false;
+			}
 		}else
 			stream->seek(size);
 	}
+	if(!uploadOk){
+		tex->destroy();
+		return nil;
+	}
 	return tex;
 }
 
-- 
2.43.0
